### PR TITLE
Archive ranges as well as doc lines

### PR DESCRIPTION
**Pull request by jpallen**

### CODE REVIEW

File: app/coffee/DocArchiveManager.coffee [hunk 1]
```diff
@@ -60,26 +66,50 @@ module.exports = DocArchive =
...
+				mongo_doc.ranges = RangeManager.jsonRangesToMongo(doc.ranges)
+		else if doc instanceof Array
+			mongo_doc.lines = doc
+		else
+			return callback(new Error("I don't understand the doc format in s3"))
+		return callback null, mongo_doc
+
+	_mongoDocToS3Doc: (doc, callback = (error, s3_doc) ->) ->
+		json = JSON.stringify({
+			lines: doc.lines
```
Comment (henryoswald) [inline app/coffee/DocArchiveManager.coffee]: do we want to null check doc.lines and doc.ranges to ensure they have been passed through Fully and correctly?

File: app/coffee/DocArchiveManager.coffee [hunk 2]
```diff
@@ -60,26 +66,50 @@ module.exports = DocArchive =
...
+			ranges: doc.ranges
```
Comment (henryoswald) [inline app/coffee/DocArchiveManager.coffee]: do we want to null check doc.lines and doc.ranges to ensure they have been passed through Fully and correctly?

File: app/coffee/DocArchiveManager.coffee [hunk 3]
```diff
@@ -26,21 +27,26 @@ module.exports = DocArchive =
...
-				logger.err responseMD5:md5response, linesMD5:md5lines,  project_id:project_id, doc_id: doc?._id, "err in response md5 from s3"
-				return callback new Error("Error in S3 md5 response")
-			MongoManager.markDocAsArchived doc._id, doc.rev, (err) ->
-				return callback(err) if err?
-				callback()
+		DocArchive._mongoDocToS3Doc doc, (error, json_doc) ->
+			return callback(error) if error?
+			options.body = json_doc
+			options.headers =
+				'Content-Type': "application/json"
```
Comment (henryoswald) [inline app/coffee/DocArchiveManager.coffee]: is this added for a reason?

Comment (henryoswald) [inline app/coffee/DocArchiveManager.coffee]: is this added for a reason?

File: app/coffee/DocArchiveManager.coffee [hunk 4]
```diff
@@ -26,21 +27,26 @@ module.exports = DocArchive =
...
+			request.put options, (err, res) ->
+				if err? || res.statusCode != 200
+					logger.err err:err, res:res, project_id:project_id, doc_id: doc._id, statusCode: res?.statusCode, "something went wrong archiving doc in aws"
+					return callback new Error("Error in S3 request")
+				md5lines = crypto.createHash("md5").update(json_doc, "utf8").digest("hex")
+				md5response = res.headers.etag.toString().replace(/\"/g, '')
+				if md5lines != md5response
+					logger.err responseMD5:md5response, linesMD5:md5lines,  project_id:project_id, doc_id: doc?._id, "err in response md5 from s3"
```
Comment (henryoswald) [inline app/coffee/DocArchiveManager.coffee]: could this func be getting a bit big. worth pulling this md5 check it into separate function? (optional)

File: app/coffee/DocArchiveManager.coffee [hunk 5]
```diff
@@ -60,26 +66,50 @@ module.exports = DocArchive =
...
+				MongoManager.upsertIntoDocCollection project_id, doc_id.toString(), mongo_doc, (err) ->
+					return callback(err) if err?
+					logger.log project_id: project_id, doc_id: doc_id, "deleting doc from s3"
+					request.del options, (err, res, body)->
+						if err? || res.statusCode != 204
+							logger.err err:err, res:res, project_id:project_id, doc_id:doc_id, "something went wrong deleting doc from aws"
+							return callback new Errors.NotFoundError("Error in S3 request")
+						callback()
+
+	_s3DocToMongoDoc: (doc, callback = (error, mongo_doc) ->) ->
```
Comment (henryoswald) [inline app/coffee/DocArchiveManager.coffee]: no unit tests around _s3DocToMongoDoc, just relying on acceptance?

File: app/coffee/DocArchiveManager.coffee (hunk 3 above)
Comment (jpallen) [inline app/coffee/DocArchiveManager.coffee]: We previously set it via `json: doc`, but we need to do it manually now to have the JSON string to compare the md5 hash to.

File: app/coffee/DocArchiveManager.coffee (hunk 2 above)
Comment (jpallen) [inline app/coffee/DocArchiveManager.coffee]: Yes, this is a good idea

File: app/coffee/DocArchiveManager.coffee (hunk 5 above)
Comment (jpallen) [inline app/coffee/DocArchiveManager.coffee]: Actually, it's not, I'll add tests

File: app/coffee/DocArchiveManager.coffee (hunk 4 above)
Comment (jpallen) [inline app/coffee/DocArchiveManager.coffee]: Will do

File: app/coffee/DocArchiveManager.coffee (hunk 3 above)
Comment (jpallen) [inline app/coffee/DocArchiveManager.coffee]: Duplicate, see above.

File: app/coffee/DocArchiveManager.coffee (hunk 4 above)
Comment (jpallen) [inline app/coffee/DocArchiveManager.coffee]: It's only used here, and is easier to test in line do the tight coupling with the consistency checking/error state, so I decided to leave it.

**Pull request merged**